docs/jax-101/02-jitting.ipynb

```diff
@@ -0,0 +1,570 @@
+     ]
+    }
+   ],
+   "source": [
+    "# While loop conditioned on x and n.\n",
+    "\n",
+    "def g(x, n):\n",
+    "  i = 0\n",
+    "  while i < n:\n",
+    "    i += 1\n",
+    "  return x + i\n",
+    "\n",
+    "g_jit = jax.jit(g)\n",
+    "g_jit(10, 20)  # Should raise an error. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "isz2U_XX_wH2"
+   },
+   "source": [
+    "As you see, running the examples above will give an error. \n",
+    "\n",
+    "The problem is that we tried to condition on the *value* of an input to the function being jitted. The reason we can't do this is related to the fact mentioned above that jaxpr depends on the actual values used to trace it. \n",
+    "\n",
+    "The more specific information about the values we use in the trace, the more we can use standard Python control flow to express ourselves. However, being too specific means we can't reuse the same traced function for other values. JAX solves this by tracing at different levels of abstraction for different purposes.\n",
+    "\n",
+    "For `jax.jit`, the default level is `ShapedArray` -- that is, each tracer has a concrete shape (which we're allowed to condition on), but no concrete value. This allows the compiled function to work on all possible inputs with the same shape -- the standard use case in machine learning. However, because the tracers have no concrete value, if we attempt to condition on one, we get the error above.\n",
+    "\n",
+    "In `jax.grad`, the constraints are more relaxed, so you can do more. If you compose several transformations, however, you must satisfy the constraints of the most strict one. So, if you `jit(grad(f))`, `f` mustn't condition on value. For more detail on the interaction between Python control flow and JAX, see [this section](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Control-Flow) of the JAX docs.\n",
+    "\n",
+    "One way to deal with this problem is to rewrite the code to avoid conditionals on value. Another is to use special [control flow operators](https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators) like `jax.lax.cond`. However, sometimes that is impossible. In that case, you can consider jitting only part of the function. For example, if the most computationally expensive part of the function is inside the loop, we can jit just that inner part (though make sure to check the next section on caching to avoid shooting yourself in the foot):"
```

Comment:
Suggestion:

- "JITting" since you wrote "to JIT" in https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html#to-jit-or-not-to-jit

```suggestion
    "One way to deal with this problem is to rewrite the code to avoid conditionals on value. Another is to use special [control flow operators](https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators) like `jax.lax.cond`. However, sometimes that is impossible. In that case, you can consider JITting only part of the function. For example, if the most computationally expensive part of the function is inside the loop, we can JIT just that inner part (though make sure to check the next section on caching to avoid shooting yourself in the foot):"
```